Add QML support and rewrite the log page

flowln opened the pull request:
Maybe they should be separated, but i don't want to keep two branches in sync :)

The rewrite is 1. a way to have QML in action, and 2. an attempt at fixing the long-standing loading issues of that page.

~First couple commits add qml support, the last ones are the rework.~ Stil follows roughly that, but some of the latter commits are intertwined.

Known issues:
- [x] Nix workflow is dead
- [x] Buttons on Windows have a weird background color
- [x] Warning for cyclic binding on the 'color' property in LogPage

Things that need to be done:
- [x] Add linebreak when copying to clipboard (easy)
- [ ] Review code (since it will probably dictate how development will flow in the future, we should take a close look at how things are structured, i'm definitely not sure the way i've done it is the best)

i will update this description in the future

flowln committed:
feat: add QML and friends support in CMakeFiles

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
refactor: move `guessLevel` to MessageLevel and optimize it

This avoids doing a lot of QRegularExpression object creations when
parsing the logs, which was one of two major bottlenecks in log processing
in the launcher.

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat(Theming): allow using custom conf file for theming QQC

It does make the UX of setting the theme a bit worse, but there's
nothing we can do, besides auto-restarting the launcher perhaps.

This uses a new file in the root directory `.qml_theme` to store the
`qtquickcontrols2.conf` file to load at startup, before any of the Qt
machinery starts. Hope this works as intended in all platforms and
configurations. :)

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat(Theming): Add ThemedControl QML Component

This sets up the correct palettes to reflect the theme colors.

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat(ui): rewrite LogPage in QML

The big thing here is using a ListView as a big TextArea, since it
prevents the application taking ages to create the QTextDocument and
laying out the text before showing it. It could be improved further by
making the delegate creating asynchronous, but my QML-fu is not at that
level just yet.

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat(actions): add QML and friends to the CI builds

(unless i messed up)

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat(build): install qml plugins on the *bundle mess*

Signed-off-by: flow <flowlnlnln@gmail.com>

Ryex committed:
packaging(macos): fix illegal dotted paths in macos bundle for codesigning

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>
Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
packaging(Windows): add QtQml and QtQuick folders to NSI installer

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat: remember the instance's Minecraft log's wrap mode

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix(macos): bundle all the qml stuff needed correctly

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat: optimize log viewer's scrolling

This makes scrolling through a lot of log lines much more responsive.

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix+feat: optimize and fix log viewer's selection

This both fixes an issue created in the previous commit with selection
inside the line (as oposed to multi-line selections), and optimizes it
to not freeze stuff when selecting a bunch of lines at once.

Now the selection model is kept small, at the loss of future flexibility
to allow non-contiguous selections, which isn't supported in the current
version anyway.

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat: add more fine-grained search result highlight in log viewer

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix: add missing QDebug include for Qt5

Signed-off-by: flow <flowlnlnln@gmail.com>

Scrumplex committed:
fix(nix): add missing QQC2 dependency for Qt 5

(Upstream changed this stuff, so I tried to follow, but I have no idea
it it works - flow)

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>
Signed-off-by: flow <flowlnlnln@gmail.com>

Scrumplex commented:
Nix works now :P

DioEgizio commented:
the custom qqc style thing doesn't work even if i have the file in the theme folder
it just tries to apply it anyways on the system theme, so the logs get spammed with:
`    24.489 W | qrc:/components/ListBasedTextArea.qml:272:29: QML Rectangle: The current style does not support customization of this control (property: "background" item: QQuickRectangle(0x6000023a2680, parent=0x0, geometry=0,0 0x0)). Please customize a non-native style (such as Basic, Fusion, Material, etc). For more information, see: https://doc.qt.io/qt-6/qtquickcontrols2-customize.html#customization-reference`

Ryex commented on launcher/MessageLevel.cpp:
```diff
@@ -34,3 +38,59 @@ MessageLevel::Enum MessageLevel::fromLine(QString &line)
+        if(levelStr == QLatin1String("FATAL"))
+            level = MessageLevel::Fatal;
+        if(levelStr == QLatin1String("TRACE") || levelStr == QLatin1String("DEBUG"))
+            level = MessageLevel::Debug;
```
these are all `if {}` instead of `if {} else if {}` so there is no short circuiting for the later checks? perhaps this is written better as a switch statement?
I know this is the original code but it seems as good a time as any to optimize

Trial97 replied:
A switch doesn't really work with the strings

Ryex commented on launcher/MessageLevel.cpp:
```diff
@@ -34,3 +38,59 @@ MessageLevel::Enum MessageLevel::fromLine(QString &line)
+        if (line.contains(QLatin1String("[WARNING]")))
+            level = MessageLevel::Warning;
+        if (line.contains(QLatin1String("[DEBUG]")))
+            level = MessageLevel::Debug;
```
same as above but using `contains` instead of equality. that's a loop for every call so it's a bit worse.

Ryex commented on launcher/ui/pages/instance/LogPage.cpp:
```diff
@@ -36,331 +36,288 @@
+    [[maybe_unused]] bool no_errors = QQmlProperty::write(rootObject(), "logModel", QVariant::fromValue(m_proxy));
+    Q_ASSERT(no_errors);
+
+    no_errors = connect(rootObject(), SIGNAL(onWrapModeChanged(int)), this, SLOT(wrapModeChanged(int)));
+    Q_ASSERT(no_errors);
```
This `Q_ASSERT(no_errors);` pattern seems like something that will need to be reused with every Qml object we want to connect. should we write a simple macro?

Ryex commented on launcher/ui/pages/instance/LogPage.cpp:
```diff
@@ -36,331 +36,288 @@
+void LogPage::wrapModeChanged(int new_state)
 {
-    return true;
+    [[maybe_unused]] bool no_errors = QQmlProperty::write(rootObject(), "wrapMode", new_state);
+    Q_ASSERT(no_errors);
```
yeah, a macro that performs that assert without the temporary stack var but retains the readability would be nice

flowln committed:
feat: optimize a bit more MessageLevel

QLatin1String is faster for QString comparisons, and we can use some
short-circuits with if...else if.

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat: add helper for QML calls from C++ that must succeed

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
meta: add test for qmllint

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix: a couple warnings given by qmllint

Signed-off-by: flow <flowlnlnln@gmail.com>

Scrumplex commented:
any activity on this? Or can we close this and hope someone else picks up the work?

flowln committed:
fix(LogPage): 'color' property cyclic binding warning

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix(LogPage): 'assing [undefined] to bool' warning

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix(LogPage): regex search not working and not doing a global match

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix(LogPage): add newlines when copying from clipboard

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
meta: add basic tests for QML theming

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix: always set QML fallback style to 'Fusion'

This prevents issues when first opening the interface, especially on
Qt5, that previously would show up with wrong colors and stuff.

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln commented:
Glad you remembered me to continue working on this :P

So, the current state of this is:

Tests will fail on Qt 5.12, due to it missing the 'required' keyword on QML. This is pretty annoying to deal with, since the presence of that keyword has a special meaning on views, which is what I'm using in the log page. ~can we bump the minimum qt version to 5.15 pwease~

I don't have any Windows or Mac machines, so testing this on other platforms is a nightmare. I was able to make the bundling work on those, but other than that, I have no idea how well or badly supported they are atm.

There are a couple little visual annoyances on the log page, that my limited qml knowledge does not allow me to solve. It does make the page loading a lot better for big logs though, so i'm satisfied with it. Please take a look and tell me what you think :-)

I think that's about it. I definitely like to attack this from time to time, but without being able to test it on other setups, my ability to finish this maybe isn't up for the task. If someone wants to pick it up, or give me feedback, feel welcome to do so. :slightly_smiling_face:

Ryex commented:
I'd be happy to tackle this on the windows end.

TheKodeToad commented:
this looks nice :D
what will qml/quick provide vs xml/widgets in this case though?

The single benefit i can see currently is animated scrolling, which would be better in every container with a toggle in settings - similar to IntelliJ (etc.)

flowln commented:
> this looks nice :D what will qml/quick provide vs xml/widgets in this case though?

Mainly fixing performance issues like #463 and #485. Since the text display widgets need to layout the entire text content before displaying, it slows down to a craw with big logs. The way I found to fix that (without throwing it into another thread and hiding it from the user until it loads, which I didn't find to be good UX)  is to use a list view to lazily layout the content, but doing that in QtWidgets seemed even more of a pain than doing this PR :P

DioEgizio commented:
> can we bump the minimum qt version to 5.15 pwease

seriously though I think it's time to bump to Qt 5.15. ~~Pr for than when~~

TheKodeToad commented:
When changing `getLogFileRoot` to `.minecraft/logs` instead of `.minecraft`, a performance issue I managed to reproduce seemed to disappear (due to less folders having to be indexed and watched).

Edit: nvm, now it's gone away without the patch (??). weird

TheKodeToad commented:
Huh. I've managed to reproduce #463, and it disappears after an application restart...

flowln commented:
> When changing `getLogFileRoot` to `.minecraft/logs` instead of `.minecraft`, a performance issue I managed to reproduce seemed to disappear (due to less folders having to be indexed and watched).

This is an optimization mostly related to `OtherLogsPage`. Indeed, there's performance gains to be had by reducing the amount of stuff done in the the instance pages constructors, and moving them to `openedImpl` instead. That does not, however, change the fact that the main log page (the one that appears when you ran the instance in that session) can get quite laggy with big logs, especially on HDDs.

> Huh. I've managed to reproduce https://github.com/PrismLauncher/PrismLauncher/issues/463, and it disappears after an application restart...

It's possible your OS cached the log file reads from disk, and so the retrieval gets done faster after the restart, though I wouldn't be able to say for sure.

What I know for sure is that the issue exists, don't worry about that :)
You can try artificially adding a lot of log lines on something like `LaunchTask::onLogLine` to simulate the effect of having a big log file to load.

TheKodeToad commented:
i was investigating #1450 :P

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into qml_support_with_logs

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
~~I tried fixing the conflicts for this, but I'm not able to test if the page works(it builds at least).~~
Able to launch and view the fabulous logs page so pushing the merge :D

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into qml_support_with_logs

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into qml_support_with_logs

Trial97 committed:
format and fix nix

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
If you get `module "kvantum" is not installed` warning and you are not able to view the logs page add: `QT_STYLE_OVERRIDE=""` before the executable

Trial97 committed:
Upgraded ubuntu image

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Added libfuse2

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
Fixed the CI:
- merged the old nix package with the new files(hope I did not break a thing)
- upgraded the Ubuntu image from 20.04 to 22.04 because the Qt5 tools were for qt5.12 instead of qt5.15(https://packages.ubuntu.com/search?suite=default&section=all&arch=any&keywords=qtdeclarative5-dev-tools&searchon=names):
![image](https://github.com/PrismLauncher/PrismLauncher/assets/42720688/8131de22-3319-4c02-b0af-fc0e21528403)
- because of the image upgrade I needed to install the `libfuse2` for appImage(https://github.com/AppImage/AppImageKit/wiki/FUSE)

The only thing remaining for this should be testing on macos, windows, and linux. So if somebody wants to test it on a specific platform just grab it from the latest actions(preferably the portable versions) and test it out.

I would have wanted to mark this draft ready for review but from the comments, I could not figure out if something needs to be done other than testing.

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into qml_support_with_logs

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Scrumplex commented:
Perhaps we can utilize this: https://api.kde.org/frameworks/qqc2-desktop-style/html/index.html

DioEgizio commented:
yeah that's pretty good. important to note that rn in the master branch [it requires Qt 6.5 or higher](https://invent.kde.org/frameworks/qqc2-desktop-style/-/blob/master/CMakeLists.txt?ref_type=heads#L8) so I guess that's most likely gonna be the requirement on stable kf6 too

DioEgizio commented:
> upgraded the Ubuntu image from 20.04 to 22.04 because the Qt5 tools were for qt5.12 instead of qt5.15(packages.ubuntu.com/search?suite=default§ion=all&arch=any&keywords=qtdeclarative5-dev-tools&searchon=names):

That's gonna break compat with older distros

Scrumplex commented:
> > upgraded the Ubuntu image from 20.04 to 22.04 because the Qt5 tools were for qt5.12 instead of qt5.15(packages.ubuntu.com/search?suite=default§ion=all&arch=any&keywords=qtdeclarative5-dev-tools&searchon=names):
> 
> That's gonna break compat with older distros

Realistically we are gonna drop Qt 5 eventually anyway

Trial97 commented:
> Perhaps we can utilize this: https://api.kde.org/frameworks/qqc2-desktop-style/html/index.html

I need to look into it.

> That's gonna break compat with older distros

I'm aware that breaks stuff but due to what this PR tries to achieve that is understandable. We would plan to remove the QT 5 sometime after QT6 is stable enough. From a few discussions on discord we may start looking into upgrading to the qt5 version used in this PR somewhere in April next year.

Trial97 commented:
> Perhaps we can utilize this: https://api.kde.org/frameworks/qqc2-desktop-style/html/index.html

How should we use this more exactly? 
As a fallback style(instead of the current `Fusion`)?
From what I see this can be controlled by an environment variable so do we need to change any code?

Trial97 commented:
~~Scrumplex as you were the one that added code signing for mac, can you please fix the conflicts on this branch?
It looks like flowln had some signing code for qml resources. I do not understand the mac stuff and I do not want to break this PR.~~ I did try to fix it so hopefully it works

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into qml_support_with_logs

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
format the LogPage

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
remove prepare_JREs.sh line

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
add qml dest directory

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
increment libffi

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
Last status on this:
- for me on qt6 the QQuickWidget::statusChanged is not triggered
- also for me the bottom row components overlap:
![image](https://github.com/user-attachments/assets/f73c2776-522e-4877-9d5e-3ebf26eb68bf)
- no idea what Scrumplex meant with "Perhaps we can utilize this: https://api.kde.org/frameworks/qqc2-desktop-style/html/index.html"
- I did not touched windows so: "Buttons on Windows have a weird background color" is still unchecked in the original todo comment
- and obviously needs reviews

And no I don't know how to fix the mentioned stuff

Ryex committed:
fixes for state, layout, theme

  - State signal not sent if state is READY from qml source set before
    constructor
  - native "Windows" theme doe snto seem tow work correctly
  - QML layout need better implcit heights and prefer width to layout
    properly

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
fixed the `QQuickWidget::statusChanged` and button overlap issues

Windows Native QQStyle is broken, temp forced the default to Fusion. I'll look into the kde desktop style solution as they may be preferred to any QQuickStyle option (all our themes are based off of QStyle which the kde desktop style defers to)

Trial97 commented:
Closing this PR: QML integration is no longer planned, and this PR is no longer active. Additionally, the number of conflicts has grown significantly, making it impractical to continue.
